src/functor.jl

```diff
@@ -169,7 +169,52 @@ function _default_walk(f, x)
   func, re = functor(x)
   re(map(f, func))
 end
-_default_walk(f, ::Nothing, ::Nothing) = nothing
+_default_walk(_, ::Nothing, ::Nothing) = nothing
+
+# Side effects only, saves a restructure
+function _foreach_walk(f, x)
+  foreach(f, children(x))
+  return x
+end
+
+### WARNING: the following is unstable internal functionality. Use at your own risk!
+# Wrapper over an IdDict which only saves values with a stable object identity
+struct Cache{K,V}
+  inner::IdDict{K,V}
+end
+Cache() = Cache(IdDict())
+
+iscachesafe(x) = !isbits(x) && ismutable(x)
+# Functionally immutable and observe value semantics, but still `ismutable` and not `isbits`
+iscachesafe(::Union{String,Symbol}) = false
+# For varargs
+iscachesafe(xs::Tuple) = all(iscachesafe, xs)
+Base.get!(f, c::Cache, x) = iscachesafe(x) ? get!(f, c.inner, x) : f(x)
+
+# Passthrough used to disable caching (e.g. when passing `cache=false`)
+struct NoCache end
+Base.get!(f, ::NoCache, x) = f(x)
+
+# Encapsulates the self-recursive part of a recursive tree reduction (fold).
+# This allows calling functions to remove any self-calls or nested callback closures.
+struct Fold{F,L,C,W}
+  fn::F
+  isleaf::L
+  cache::C
+  walk::W
+end
+(fld::Fold)(x) = get!(() -> fld.fn(fld.isleaf(x) ? x : fld.walk(fld, x)), fld.cache, x)
+
+# Convenience function for working with `Fold`
+function fold(f, x; isleaf = isleaf, cache = false, walk = _default_walk)
+  if cache === true
```

Comment:
Note that `fmap` and `fcollect` are still passing through `IdDict()` by default. The hope is that new functions either don't allow customization of the cache or use a "safer" default.